packages/blocks/src/root-block/edgeless/components/toolbar/shape/shape-draggable.ts

```diff
@@ -0,0 +1,242 @@
+
+  @property({ attribute: false })
+  accessor color!: string;
+
+  @property({ attribute: false })
+  accessor stroke!: string;
+
+  @property({ attribute: false })
+  accessor onShapeClick: (shape: DraggableShape) => void = () => {};
+
+  public initDragController() {
+    if (!this.edgeless || !this.toolbarContainer) return;
+    if (this.draggableController) return;
+    this.draggableController = new EdgelessDraggableElementController(this, {
+      service: this.edgeless.service,
+      edgeless: this.edgeless,
+      scopeElement: this.toolbarContainer,
+      standardWidth: 100,
+      onOverlayCraeted: overlay => {
+        Object.assign(overlay.element.style, {
+          color: this.color,
+          stroke: this.stroke,
+        });
+        const controller = this.edgeless.tools.currentController;
+        if (controller instanceof ShapeToolController) {
+          controller.clearOverlay();
+        }
+      },
+      onDrop: (el, bound) => {
+        const xywh = bound.serialize();
+        const shape = el.data;
+        this.edgeless.service.addElement(CanvasElementType.SHAPE, {
+          shapeType: shape.name === 'roundedRect' ? ShapeType.Rect : shape.name,
+          xywh,
+          radius: shape.name === 'roundedRect' ? 0.1 : 0,
+        });
+      },
+      onElementClick: el => {
+        this.onShapeClick?.(el.data);
+      },
+      onEnterOrLeaveScope: (overlay, isOutside) => {
+        overlay.element.style.filter = isOutside
+          ? 'none'
+          : `drop-shadow(${this.shapeShadow})`;
+      },
+    });
+  }
+
+  override updated(_changedProperties: Map<PropertyKey, unknown>) {
+    const controllerRequiredProps = ['edgeless', 'toolbarContainer'] as const;
+    if (
+      controllerRequiredProps.some(p => _changedProperties.has(p)) &&
+      !this.draggableController
+    ) {
+      this.initDragController();
+    }
+  }
+
+  override render() {
+    const { cancelled, dragOut, draggingElement } =
+      this.draggableController?.states || {};
+    const draggingShape = draggingElement?.data;
+    return html`<div class="edgeless-shape-draggable">
+      ${repeat(
+        shapes,
+        s => s.name,
+        shape => {
+          const isBeingDragged = draggingShape?.name === shape.name;
+          const baseStyle = {
+            ...buildVariablesObject(shape.style),
+            filter: `drop-shadow(${this.shapeShadow})`,
+          };
+          const currStyle = styleMap({
+            ...baseStyle,
+            opacity: isBeingDragged ? 0 : 1,
+            pointerEvents: draggingShape ? 'none' : 'auto',
+          });
+          const nextStyle = styleMap(baseStyle);
+          return html`${isBeingDragged
+              ? html`<div
+                  style=${nextStyle}
+                  class=${classMap({
+                    shape: true,
+                    next: true,
+                    coming: !!dragOut && !cancelled,
+                  })}
+                >
+                  ${shape.svg}
```

Comment:
I think, we can use [clip-path](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path) to clip region.

https://github.com/toeverything/blocksuite/assets/27926/55366816-bae5-4f51-ba2b-cd40f1e91d0e

Reply:
Not quite clear your purpose, in my opinion:
- we have to maintain the path for each shape if using clip-path
- the pointer trigger region has been handled here
- the wrapper's height not matching the svg now, because the svg is displayed `inline`, I'll fix that later.

Reply:
When the mouse moves out of the triangle,  the rectangle should be hovered.

Reply:
Yes that should have already been handled. I've fixed that now. We are using css only to achieve triggering pointer events on shape rather than shape bounding rect:
```css
.wrapper {
    pointer-events: none;
}

.wrapper path,
.wrapper circle,
.wrapper rect {
    pointer-events: auto;
}
```